handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1447,65 +1484,224 @@ private static ByteBuffer toByteBuffer(ByteBuf out, int index, int len) {
                 out.nioBuffer(index, len);
     }
 
+    private static boolean inEventLoop(Executor executor) {
+        return executor instanceof EventExecutor && ((EventExecutor) executor).inEventLoop();
+    }
+
     /**
-     * Fetches all delegated tasks from the {@link SSLEngine} and runs them via the {@link #delegatedTaskExecutor}.
-     * If the {@link #delegatedTaskExecutor} is {@link ImmediateExecutor}, just call {@link Runnable#run()} directly
-     * instead of using {@link Executor#execute(Runnable)}.  Otherwise, run the tasks via
-     * the {@link #delegatedTaskExecutor} and wait until the tasks are finished.
+     * Will either run the delegating task directly calling {@link Runnable#run()} and return {@code true} or will
+     * offload the delegating task using {@link Executor#execute(Runnable)} and return {@code false}.
+     *
+     * If the task is offloaded it will take care to resume its work on the {@link EventExecutor} once there are no
+     * more tasks to process.
      */
-    private void runDelegatedTasks() {
-        if (delegatedTaskExecutor == ImmediateExecutor.INSTANCE) {
+    private boolean runDelegatedTasks(boolean inUnwrap) {
+        if (delegatedTaskExecutor == ImmediateExecutor.INSTANCE || inEventLoop(delegatedTaskExecutor)) {
+            // We should run the task directly in the EventExecutor thread and not offload at all.
             for (;;) {
                 Runnable task = engine.getDelegatedTask();
                 if (task == null) {
-                    break;
+                    return true;
                 }
-
                 task.run();
             }
         } else {
-            final List<Runnable> tasks = new ArrayList<Runnable>(2);
-            for (;;) {
-                final Runnable task = engine.getDelegatedTask();
-                if (task == null) {
-                    break;
+            final Runnable task = engine.getDelegatedTask();
+            // We only call the method when NEED_TASK is returned so we must have a task!
+            assert task != null;
+            executeDelegatingTask(task, inUnwrap);
```

Comment:
Could you avoid passing an explicit task to the executor here? I think a constant `Runnable` would suffice, i.e. no need to allocate `SslTask`s - call `engine.getDelegatedTask()` first thing in the `run()` method. `inUnwrap` could be a plain instance field.

Reply:
I think allocate `SslTask`s is cleaner and as this is only during handshake its fine.

Reply:
@normanmaurer good point, probably better to avoid the extra fields on the handler since they only apply during handshake. Kind of tangential but do you think it could be worth grouping all the handshake-related fields (including the executor) into a separate object which can be discarded once the handshake is complete? Just an idea, maybe limited benefit :)

It would still be a simple change I think to remove the `task` field from `SslTask` and have it grab the task itself from the engine. This eleminates a synchronized method call from the event loop and the next task dispatch in the `run()` method could be changed to `delegatedTaskExecutor.execute(this)`. wdyt?

Also given the current logic I don't think this can happen, but in general `NEED_TASK` doesn't mean that `getDelegatedTask()` won't return null. It remains in that state until previously returned task(s) have finished running. Might it be possible to end up in this situation if the delegated executor rejects the task?

Reply:
thats a good point... maybe I should better remove the assert and just return if task == null to be safe in the future.